src/workspace/mod.rs

```diff
@@ -0,0 +1,1823 @@
+            opts.fs,
+            opts.pkg_json_cache,
+          );
+          match pkg_json_result {
+            Ok(pkg_json) => Ok(DenoOrPkgJson::PkgJson(pkg_json)),
+            Err(PackageJsonLoadError::Io { source, .. })
+              if source.kind() == std::io::ErrorKind::NotFound =>
+            {
+              Err(ResolveWorkspaceMemberError::NotFound {
+                dir_url: member_dir_url.clone(),
+              })
+            }
+            Err(err) => Err(err.into()),
+          }
+        };
+
+        let member = if !raw_member.ends_with('/') {
+          Cow::Owned(format!("{}/", raw_member))
+        } else {
+          Cow::Borrowed(raw_member)
+        };
+        let member_dir_url = root_config_file_directory_url
+          .join(&member)
+          .map_err(|err| {
+            WorkspaceDiscoverError(
+              WorkspaceDiscoverErrorKind::InvalidMember {
+                base: workspace_config_file.specifier.clone(),
+                member: raw_member.clone(),
+                source: err,
+              }
+              .into(),
+            )
+          })?;
+        if !member_dir_url
+          .as_str()
+          .starts_with(root_config_file_directory_url.as_str())
+        {
+          return Err(
+            ResolveWorkspaceMemberError::NonDescendant {
+              workspace_url: root_config_file_directory_url.clone(),
+              member_url: member_dir_url,
+            }
+            .into(),
+          );
+        }
+        let config = find_config(&member_dir_url)?;
+        final_members.insert(new_rc(member_dir_url), config);
+      }
+      if let Some(config_url) = found_configs.into_keys().next() {
+        return Err(
+          WorkspaceDiscoverErrorKind::ConfigNotWorkspaceMember {
+            workspace_url: workspace_config_file.specifier.clone(),
+            config_url,
+          }
+          .into(),
+        );
+      }
+      return Ok(ConfigFileDiscovery::Workspace {
+        root: new_rc(workspace_config_file),
+        members: final_members,
+      });
+    } else {
+      if first_config_file.is_none() {
+        first_config_file = Some(workspace_config_file.specifier.clone());
+      }
+      next_start_dir =
+        specifier_to_file_path(&workspace_config_file.specifier)?
+          .parent()
+          .and_then(|p| p.parent())
+          .map(|p| Cow::Owned(p.to_owned()));
+      found_configs.insert(
+        workspace_config_file.specifier.clone(),
+        workspace_config_file,
+      );
+    }
+  }
+
+  if let Some(first_config_file) = first_config_file {
+    let config = found_configs.remove(&first_config_file).unwrap();
+    Ok(ConfigFileDiscovery::Single(new_rc(config)))
+  } else {
+    Ok(ConfigFileDiscovery::None)
+  }
+}
+
+#[derive(Debug)]
+enum PackageJsonDiscovery {
+  None,
+  Single(PackageJsonRc),
+  Workspace {
+    root: PackageJsonRc,
+    members: BTreeMap<UrlRc, PackageJsonRc>,
+  },
+}
+
+fn discover_with_npm(
```

Comment:
These discovery algos are massive - should we move them to `src/workspace/discovery.rs`?